### DISCUSSION TITLE
Additional Attributes and Services for better Lovelace Cards

### DISCUSSION THREAD

**Discussion by rxritalin** [music-assistant/support#2586] [Feature requests and ideas]
I originally posted this in Q&A section, but I dont think what I am looking to do is possible yet with MASS, so I am reposting here in the Feature Requests. 

**Whats already possible**
I am pretty much finished or have gone as far as I can go at present with a full screen card.
This UI is intended to be a single player, on a single device and act as the main UI to interface with MASS. Its one of those odd 1920x720 touch displays, incase you are wondering why its so wide.
At present I am using the picture elements card as my base and "mini-media-player" card for most of the interactive elements, as it seems to offer the most flexibility. The layout is obviously based on Volumio's UI with a few changes to make it more touch friendly and easier to read at a distance.

![MASS - Screen 1](https://github.com/music-assistant/support/assets/3268460/69465fb3-1759-4f8a-8be9-e02aa9056bf3)
![MASS - Screen 2](https://github.com/music-assistant/support/assets/3268460/f2c2694d-3a77-4b4f-8543-03444a072528)

I can obviously already use the core features of a media card;
cover art,
track
title,
album
playtime/runtime
play/pause, next, prev, shuffle and repeat
volume

I have also found a simple enough way to access the HASS media player options for Search, PlayLists, Artist, etc via popups housing an iFrame Card. This works but its not ideal as each pages needs to load in the iFrame each time its accessed. It makes the UI feel a bit clunky but at least it works. 
This is about as far as I can push it right now. 
![MASS - Screen 4](https://github.com/music-assistant/support/assets/3268460/76497347-661b-462f-936a-237aa647c2d8)

**New Features Request**
1. Missing Services 

- Add _service.squeezebox.join_ to MASS so it can utilize the built in speaker grouping feature found in Mini-Media-Card. Or create some 
   alternate accessible service to do the same and let us build our own interface. ✅(edit: OzGav Use [media_player.join](https://www.home-assistant.io/integrations/media_player/#action-media_playerjoin)
- Add a service to allow for adding a current track to favorites. [✅](https://github.com/orgs/music-assistant/discussions/2586#discussioncomment-12270724)
- Add a service to allow for adding current track to a playlist. Not exactly sure how a we could utilize this in the lovelace UI, as it would 
   require some sort of secondary popup to list out which playlist to it the tract to. [✅](https://github.com/orgs/music-assistant/discussions/2586#discussioncomment-12301977)

2. Missing Attributes

- Add an attribute that lists the name of the current playlist. The current attribute of "active_queue" outputs e0:51:d8:10:95:84, which as far 
   as I can tell is useless to the end user. If I am wrong please correct me. ❌(edit: OzGav The queue doesn't have a name)
- Add an attribute that lists the name of the current service provider (spotify, apple, quobuz) ✅(edit: OzGav Available via `get_queue` in the `stream_details.provider` key )
- Add an attribute that lists the current tracks bitrate ✅(edit: OzGav Available via the [item_by_uri API call](https://github.com/orgs/music-assistant/discussions/2586#discussioncomment-12396343)
- Add an attribute that lists the current tracks file container (ogg, flac) ✅(edit: OzGav Available via `get_queue` in the `stream_details.content_type` key )

3.  Missing Features
     Im not sure how this should best be handled, as these are not straight forward services or attributes, I believe they are a mix of the two. 

- Add some way of accessing the current playlist. ✅(edit: OzGav Available via the [player_queues/items API call](https://github.com/orgs/music-assistant/discussions/2586#discussioncomment-12389116)
   (dirty) Make the playlist popup in the MASS UI accessible directly via a URL, so that users can utilize an iFrame card to display and 
   interact with the current playlist. 
   (clean) Take the same route that "ytube_music_player" takes, which requires the user to create helper "input_selects". One would likely be 
   required for each speaker. Once the input select is created with a specific name/syntax, MASS fills in the data. 
   https://github.com/KoljaWindeler/ytube_music_player 
   Suggested naming structure: 
   input_select.mass_speakername_current_playlist, 
- Add a "clean"  way to browse through, playlists, albums, artists, tracks and stations inside of the lovelace UI. Again, you could take the 
   route that "youtube_music_player" did and have the end user create helper input selects, and have MASS fill in the data. ✅(edit: OzGav Best idea is to continue using the iframe with the [frameless option](https://www.music-assistant.io/faq/how-to/#access-the-ma-views-directly-via-url) )
   input_select.mass_playlists
   input_select.mass_albums
   input_select.mass_artists
   input_select.mass_tracks
   input_select.mass_stations. 
- Add the ability to search directly from the lovelace UI without jumping through hoops. Being able to use something like "seach_card" 
   directly from the lovelace UI would be greate. https://github.com/postlund/search-card
   If its not possible to utilize this card in its current form, perhaps creating one specific to the task. ✅(edit: OzGav Custom card: https://github.com/fastxl2024/mass-search-card )

I would love to be able to end up with something like this to finally finish off this card. 
![MASS - Screen 3](https://github.com/music-assistant/support/assets/3268460/7d038627-1a57-44e7-873b-7fa95021fc54)

**Comment by Paul-Vdp** [comment]
If this is what it takes to be able to build the perfect Mass UI, then I'm all for it. And so should everybody ...

**Reply by wambs** [reply]
This is a really good start!!! Looking forward to using this!!

**Comment by marcelveldt** [comment]
MA is never playing a playlist but a queue of items. So there is never a single playlist because the tracks on the queue could be originated from everywhere. The best you could probably do is maybe remember the last X entries from the play_media request. So you at least know which albums/playlists/artists are added

**Comment by thekiwismarthome** [comment]
This is beautiful, how do I get my hands on the yaml? please

**Comment by chezpaul** [comment]
Did you ever release this anywhere? Looks really nice.

**Comment by tylerlucas** [comment]
@rxritalin did you get support on your requests? If MASS wants wide adoption then it undoubtedly needs good support from within HA, which is severely lacking right now.

**Reply by marcelveldt** [reply]
Please enlighten me, what is severe lacking? All base functions are existing today. Note that we are a super small team so feature requests are carefully weighted on enough interest. For now all development focus goes to bug fixing so we have a stable foundation.

**Reply by tylerlucas** [reply]
What support is there today? If there is any way to get Music Assistant visibility/control into Home Assistant, it is not clearly documented or, I'd venture to guess, widely known.

Searching Google for "music assistant lovelace home assistant" gives these top results:
1. A link to a bunch of MA feature requests which don't appear to provide solutions - e.g.: this post, and [Link](https://github.com/orgs/music-assistant/discussions/1693)
2. A bunch of Reddit threads with people either: (1) asking how to integrate MA into Home Assistant, and; (2) people linking to _this_ GitHub request
3. Links to the MA site, but I'm not seeing any instructions/info on how to actually add it to Home Assistant

TL;DR: If there is a way to add MA into HA, users are not understanding how.

Considering that this particular GitHub request is so widely shared/seen - I'm curious if any support was given to @rxritalin. He's clearly invested in adding support into HA, so it would great to see him get what he needs if/when the team has free cycles. Just my .02!

**Reply by OzGav** [reply]
So just to summarise support for HA is “severely lacking“ because there isn’t a Lovelace card despite the fact that:
* You can use any of the existing media player card to control the MA players
* MA can play to HA players
* Four actions are available to play, search, make announcements and transfer queues between players
* HA announcements pause and resume the currently playing music
* Voice/Assist control is available
* The MA project worked with HA to include more of the media player actions as voice controllable in core
* We provide examples of how to include all of the other media player actions as custom sentences

Looking forward to your contributions to fill the gaps.

**Reply by tylerlucas** [reply]
@OzGav @marcelveldt apologies if my initial comment was too harsh. I am a huge fan of what you've built with Music Assistant and am excited about it. I was trying to point out that many users (including myself) do not know how to use Music Assistant from within Home Assistant. 

Is there any documentation on how to use MA within HA with existing cards, like the Media Player Card that you mention? Please point me. I've used HA pretty extensively for years, but haven't ever used a Media Player card -- I'm guessing there are others like me out there, hence why people are looking for solutions on Reddit etc.

Getting a custom Lovelace card like @rxritalin was attempting would be great and help adoption. Understood that there aren't resources to work on it right now.

**Reply by OzGav** [reply]
You don’t need anything clever to use a media player card. Add the card to your config and use the MA media player entity.

If you want to get clever then you could do this https://github.com/orgs/music-assistant/discussions/954#discussioncomment-10793120

Of course the whole MA UI is available via ingress (HAOS users) or iframe (docker users)

Then there are custom media players cards like

https://github.com/kalkih/mini-media-player
https://community.home-assistant.io/t/maxi-media-player/705007

**Reply by tylerlucas** [reply]
Awesome, thanks for the suggestions!

**Comment by chezpaul** [comment]
I think the integration is great. What is missing is a really nice lovelace card. Most of us. At least all the people I get inspired from their dashboards will all use an integration as a card on their dashboard. And most of us have their menu's hidden. So we need a lovelace card that looks as cool if not way better than what's out there right now. I have had Music Assistant integrated in my HA setup for as along as I can remember (many updates) but I have NEVER used it because I can't access it yet. I'm still waiting for the lovelace card. At least that's how I see it. Great work on everything working together though. You guys make it so easy to add your providers and your speakers and it all works. But not on my dashboard. Thanks for everything though. I don't want to sound like I'm dissing you guys. Not at all.

**Comment by thekiwismarthome** [comment]
I use Music Assistant all the time at home and with voice, I am part of an awesome project called View Assist  ([Git ](https://github.com/dinki/View-Assist)& [Wiki](https://dinki.github.io/View-Assist/)) which also uses it.
What I would love is a beautiful very usable intuitive card like the above for manual interaction, more for the kids and wife, the wife struggles with voice commands on all platforms.
If we could get the current building blocks for the above UI, I am sure the community can help sort the rest out.

**Comment by chezpaul** [comment]
I also think it might be different kinds of people who make great UI's than those who make the backend/programming of a great integration.
Music Assistant is a great product. What rxritalin is working on could become a great UI for Music Assistant. And once released, I bet way more people will jump onto using Music assistant.

Why not try to implement some of rxritalin's ideas in order to make the UI work as it should and hence make Music Assistant incredible?

**Reply by OzGav** [reply]
> Why not try to implement some of rxritalin's ideas

It’s not a matter of trying, that suggests we are refusing to help him, its workload. This is OPEN SOURCE so anyone can contribute.

**Reply by chezpaul** [reply]
Yeah, I didn't even think of that. My bad. ha!
Well I hope someone at some point will have fun with it. 🤷🏼‍♂️
I'll be waiting on the sidelines rooting for Music assistant. It makes dealing with music so much easier in Home Assistant.
Thanks to all that work on this project. 🙏🏻 UI or not. (and by UI, I mean Lovelace card advancement)

**Comment by ragaimeena** [comment]
This is exactly what I am looking for
Please share the code
Please let us know if a financial appreciation of your time and effort would help
Thank you

**Reply by chezpaul** [reply]
Yeah, I'm also willing to also show my appreciation in currency. I feel like I've been waiting months when it's only been a week or so. 
😅

**Comment by OzGav** [comment]
@rxritalin have a look at the new actions available in integration version 2024.11.0 and let us know if any of those are useful for you. Your various requests are noted and will be looked at in due course.

**Reply by OzGav** [reply]
I think you will find `get_queue` implements these requests 

- Add an attribute that lists the name of the current service provider (spotify, apple, quobuz)
- Add an attribute that lists the current tracks bitrate
- Add an attribute that lists the current tracks file container (ogg, flac)

Also in your missing services you ask for the equivalent of `service.squeezebox.join` but what about `media_player.join`?

**Reply by ragaimeena** [reply]
I am not sure if he is till looking into this. It looks like he has not responded in a while. I hope someone with similar knowledge can help us with developing this. Thank you

**Reply by jonasjeeliasson** [reply]
Not sure how to use these actions to display things easily on a Lovelace dashboard. Would think we need some sort of sensors or other components housing these values as attributes rather than calling actions/services. Maybe some sort of template sensor can be created based on the result of one of these action calls, but then that template needs to be kept up-to-date some how 🤷

**Reply by OzGav** [reply]
If you don’t have a better trigger then a template updated on a time  period using `homeassistant.update_entity` would work

**Reply by Thyraz** [reply]
People that ended in this discussion might also be interested in this topic about what's already possible for a feature-rich Lovelace integration and what more might be needed:
https://discord.com/channels/753947050995089438/1305541265924034570

**Reply by felipetoffolo1** [reply]
@Thyraz that link did not work for me. I'm not very familiar with discord. Do I have to join some server first?

**Reply by thekiwismarthome** [reply]
@felipetoffolo1 look for the Music Assistant server

**Comment by OzGav** [comment]
A REST API is available in MA that we are yet to fully document. However, in regards to being able to favourite and unfavourite in HA you can do the following. It is assuming you wish to favourite (or unfavourite) the song currently playing on `media_player.ma_picoreplayer`

```
rest_command:
  ma_add_favourite:
    url: http://localhost:8095/api
    method: POST
    headers:
      accept: "application/json, text/html"
    payload: >
      {
        "message_id": "1",
        "command": "music/favorites/add_item",
        "args": {
          "item": "{{ states.media_player.ma_picoreplayer.attributes.media_content_id }}"
        }
      }
    content_type:  'application/json; charset=utf-8'
    
  ma_remove_favourite:
    url: http://localhost:8095/api
    method: POST
    headers:
      accept: "application/json, text/html"
    payload: >
      {
        "message_id": "1",
        "command": "music/favorites/remove_item",
        "args": {
          "media_type": "track",
          "library_item_id": "{{ states.media_player.ma_picoreplayer.attributes.media_content_id.split("track/")[1] }}"
        }
      }
    content_type:  'application/json; charset=utf-8'
```

You are then able to call these commands by triggering on a button change (for example) and then simply use:

```
  actions:
    - action: rest_command.ma_add_favourite
```

**Reply by OzGav** [reply]
Continuing on. This is the `rest_command` to add a track to a playlist
```

  ma_add_playlist_tracks:
    url: http://localhost:8095/api
    method: POST
    headers:
      accept: "application/json, text/html"
    payload: >
      {
        "message_id": "1",
        "command": "music/playlists/add_playlist_tracks",
        "args": {
          "db_playlist_id": "26",
          "uris": ["library://track/3951"]
        }
      }
    content_type:  'application/json; charset=utf-8'
```

**Reply by OzGav** [reply]
This will get all the items in the queue. 

`queue_id` will be the same as the `player_id` unless the player is grouped. To confirm create a `rest_command` that calls `player_queues/all` and review the information returned. The `player_id` can be found at the top of the player settings page in MA.

The limit defaults to 500 if you omit it. You are cautioned to not set a value greater then 500 to avoid breaking your system. The practical limit will depend on the resources available on your host. 

```
rest_command:
  ma_queue_test:
    url: http://localhost:8095/api
    method: POST
    headers:
      accept: "application/json, text/html"
    payload: >
      {
        "message_id": "1",
        "command": "player_queues/items",
        "args": {
          "queue_id": "b8:27:eb:8a:b8:8e",
          "limit": 500,
          "offset": 0
        }
      }
    content_type:  'application/json; charset=utf-8'
```

**Reply by OzGav** [reply]
This will get more information about an item. The URI can be found in the `media_content_id` attribute of the player

```
  ma_get_other_item_details:
    url: "http://localhost:8095/api"
    method: POST
    headers:
      accept: "application/json, text/html"
      content_type: "application/json; charset=utf-8"
    payload: >
      {
        "message_id": "1",
        "command": "music/item_by_uri",
        "args": {
          "uri": "library://track/468"
        }
      }   
```

**Reply by joopdo** [reply]
This works! For anyone else reading this, you put the Player ID: in the queue_id.
> rest_command:
>   ma_queue_test:
>     url: http://localhost:8095/api
>     method: POST
>     headers:
>       accept: "application/json, text/html"
>     payload: >
>       {
>         "message_id": "1",
>         "command": "player_queues/items",
>         "args": {
>           "queue_id": "media_player.livingroom_radio_player",
>           "limit": 500,
>           "offset": 0
>         }
>       }
>     content_type:  'application/json; charset=utf-8'

**Comment by Huseriato** [comment]
Did this ever have been released? I'm really looking for something like this.
